flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/ExecNodeConfiguration.java

```diff
@@ -0,0 +1,108 @@
+ * ExecNodeBase#getPersistedConfig()} configuration. The persisted configuration of the {@link
+ * ExecNode} which is deserialized from the JSON plan has precedence over the {@link
+ * PlannerConfiguration}.
+ */
+@Internal
+public class ExecNodeConfiguration implements ReadableConfig {
+
+    private final ReadableConfig plannerConfig;
+
+    // See https://issues.apache.org/jira/browse/FLINK-26190
+    private final TableConfig tableConfig;
+
+    private final ReadableConfig persistedConfig;
+
+    ExecNodeConfiguration(
+            ReadableConfig plannerConfig, TableConfig tableConfig, ReadableConfig persistedConfig) {
+        this.plannerConfig = plannerConfig;
+        this.persistedConfig = persistedConfig;
+        this.tableConfig = TableConfig.getDefault();
+        this.tableConfig.setNullCheck(tableConfig.getNullCheck());
+        this.tableConfig.setDecimalContext(tableConfig.getDecimalContext());
+        this.tableConfig.addConfiguration(tableConfig.getConfiguration());
+        this.tableConfig.addConfiguration((Configuration) persistedConfig);
+    }
+
+    /**
+     * Return the merged {@link TableConfig} from {@link PlannerBase#getTableConfig()} and {@link
+     * ExecNodeBase#getPersistedConfig()}.
+     *
+     * @return the {@link TableConfig}.
+     * @deprecated This method is used only for {@link CodeGeneratorContext} and related methods,
+     *     which end up passing the {@link TableConfig} to the {@link CodeGeneratorContext}. It
+     *     should be removed once {@link CodeGeneratorContext#nullCheck()} is removed, since for all
+     *     other usages it's possible to use the {@link ReadableConfig}.
+     */
+    // See https://issues.apache.org/jira/browse/FLINK-26190
+    @Deprecated
+    public TableConfig getTableConfig() {
+        return tableConfig;
+    }
+
+    @Override
+    public <T> T get(ConfigOption<T> option) {
+        return persistedConfig.getOptional(option).orElseGet(() -> plannerConfig.get(option));
+    }
+
+    @Override
+    public <T> Optional<T> getOptional(ConfigOption<T> option) {
+        final Optional<T> tableValue = persistedConfig.getOptional(option);
+        if (tableValue.isPresent()) {
+            return tableValue;
+        }
+        return plannerConfig.getOptional(option);
+    }
+
+    /** @return The duration until state which was not updated will be retained. */
+    public long getIdleStateRetentionTime() {
+        return get(ExecutionConfigOptions.IDLE_STATE_RETENTION).toMillis();
+    }
+
+    // See https://issues.apache.org/jira/browse/FLINK-26190
+    /** See {@link TableConfig#getMaxGeneratedCodeLength()}. */
+    @Deprecated
+    public long getMaxIdleStateRetentionTime() {
+        return tableConfig.getMaxGeneratedCodeLength();
+    }
+
+    // See https://issues.apache.org/jira/browse/FLINK-26190
+    /** See {@link TableConfig#getLocalTimeZone()}. */
+    public ZoneId getLocalTimeZone() {
+        return tableConfig.getLocalTimeZone();
```

Comment:
Not to do in this PR, but does it makes sense to have in the config option builder the support for the `ZoneId` type directly, so you remove the intermediate parsing code in `TableConfig`?

Reply:
makes sense yes.